File: day2/task2.py
def is_safe(r: list[int]) -> bool:
    if len(r) <= 1:
        return True
    if len(r) == 2:
        return 1 <= abs(r[0] - r[1]) <= 3

    for x1, x2, x3 in zip(r, r[1:], r[2:]):
        if not (
            (x1 < x2 < x3 or x1 > x2 > x3)
            and 1 <= abs(x1 - x2) <= 3
            and 1 <= abs(x2 - x3) <= 3
        ):
            return False
    return True


with open("input.txt") as f:
    reports = [list(map(int, l.split(" "))) for l in f.readlines()]

res = 0
for r in reports:
    for i in range(len(r)):
        if is_safe(r[:i] + r[i + 1 :]):
            res += 1
            break

print(res)

File: day2/task2.hs
readline :: String -> [Int]
readline "" = []
readline s =
  let (n, s') = head (reads s :: [(Int, String)])
   in (n : readline s')

readlines :: [String] -> [[Int]]
readlines = foldr ((:) . readline) []

cond :: Int -> Int -> Bool
cond a b = (\x -> x >= 1 && x <= 3) (abs (a - b))

safe :: [Int] -> Bool
safe [] = True
safe [_] = True
safe [x1, x2] = cond x1 x2
safe (x1 : x2 : x3 : xs) =
  (x1 < x2 && x2 < x3 || x1 > x2 && x2 > x3)
    && cond x1 x2
    && cond x2 x3
    && safe (x2 : x3 : xs)

removeOne :: [Int] -> [Int] -> Int -> [Int]
removeOne [] _ _ = []
removeOne (x : xs) (i : is) n =
  if i == n
    then xs
    else x : removeOne xs is n

options :: [Int] -> [[Int]]
options xs = foldl (\acc n -> removeOne xs [0 ..] n : acc) [] [0 .. length xs]

main = do
  contents <- readFile "input.txt"
  let l = readlines $ lines contents
  print $
    foldl
      ( \acc xs ->
          if foldl (\acc' xs' -> acc' || safe xs') False $ options xs
            then acc + 1
            else acc
      )
      0
      l
